protocol/contracts/zaps/ZapBase.sol

```diff
@@ -0,0 +1,103 @@
+pragma solidity ^0.8.4;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+abstract contract ZapBase is Ownable {
+  using SafeERC20 for IERC20;
+
+  bool public paused;
+
+  uint256 internal constant DEADLINE = 0xf000000000000000000000000000000000000000000000000000000000000000;
+
+  // fromToken => swapTarget (per curve, univ2 and balancer) approval status
+  mapping(address => mapping(address => bool)) public approvedTargets;
+
+  receive() external payable {
+    require(msg.sender != tx.origin, "Do not send ETH directly");
+  }
+
+  /**
+    @dev Adds or removes an approved swapTarget
+    * swapTargets should be Zaps and must not be tokens!
+    @param _tokens An array of tokens
+    @param _targets An array of addresses of approved swapTargets
+    @param _isApproved An array of booleans if target is approved or not
+    */
+  function setApprovedTargets(
+    address[] calldata _tokens,
+    address[] calldata _targets,
+    bool[] calldata _isApproved
+  ) external onlyOwner {
+    uint256 _length = _isApproved.length;
+    require(_targets.length == _length && _tokens.length == _length, "Invalid Input length");
+
+    for (uint256 i = 0; i < _length; i++) {
```

Comment:
If the existing `approvedTargets` (before this is called) is larger then the new length, then there'll be some stale/residual approved targets I think?

Or is the expectation that the caller sets to these to `_isApproved=false` for the remaining? Bit of a gotcha

Reply:
not sure if I understand you completely. but since it's a mapping, default for a `Map[A][B]` pair is `false`. so the new incoming approvals can be of any length and map values will either be updated/reset or new entries set.

Reply:
At t0:
```
setApprovedTargets(
    [FRAX, TEMPLE, UNI, FXS],
    [ZEROEX, TEMPLE_AMM, TEMPLE_AMM, ZEROEX],
    [true, true, true, true]
);
```

At t1:
```
setApprovedTargets(
    [cvxFxs, TEMPLE],
    [ZEROEX, TEMPLE_AMM],
    [true, true]
);
```

I think in this case the UNI, FXS are still left as approved targets?

Reply:
I'd argue that is how we'd want a setter to work? If I don't explicitly pass in a param, I'd be surprised if the setter modified that param?

I'm a real stickler for no side effects in setters; can change the name if that side effect is expected though

Reply:
Perhaps just a naming thing to me then - when I first read 'setApprovedTargets()', and that it takes an array, I think it implies that's the full set of targets - anything else is disabled

Reply:
I'm with scoop's arguments. if I don't pass in a token/target pair, the old value should remain as is in storage